### PR TITLE
Support for undocumented v99x8 feature (3rd bit of register 1)

### PR DESCRIPTION
**Pull request by ducasp**
This implements the request #1091  , supporting the 3rd bit of VDP register 1 to change the clock source of blink.

### CODE REVIEW

File: src/video/VDP.hh [hunk 1]
```diff
@@ -372,6 +372,31 @@ public:
 		       (!blinkState << 8);
 	}
 
+	/** Expresses the state of even/odd page interchange in a mask
+	  * on the line number. If even/odd interchange is active, for some
+	  * frames lines 256..511 (page 1) are replaced by 0..255 (page 0)
+	  * and 768..1023 (page 3, if appicable) by 512..767 (page 2).
```
Comment (D15C0DE) [inline src/video/VDP.hh]: applicable

**Comment by m9710797** [comment]
@ducasp: Thanks a lot for your patch!
I didn't have time yet to review it in detail (and I'm afraid I also won't have time in the following days). But I did have a quick look during lunch break. (Because I only took a quick look my following comments may be completely wrong).

Seems like a pretty straight-forward implementation. But I'm wondering about two things:

1) Is it needed to explicitly store the blinkStateMatrix[] array? Or could it's content be calculated on demand in the getEvenOddMask() function? At first sight i *think* it should be sufficient to store a reference line number (e.g. some line at which blinkCount==0 and blinkState=false) and from that you can deduce the full content of the array. This new state (either blinkStateMatrix[] or the reference line) should become part of the savestate (via VDP::serialize()).

2) After the previous point is addressed, is it still needed to put a synchronization point on every line? Currently it's still needed to call renderer->updateBlinkState(), but I wonder if that's really required, or if it can be handled in a different way (e.g. in SDLRasterizer<Pixel>::drawDisplay() you already query the state per line). The reason why I ask is because frequent synchronization points are relatively expensive to emulate (by itself a sync point or the code that it triggers is not expensive, but it does prevent optimizations in other sub-systems).

File: src/video/VDP.hh (hunk 1 above)
Comment (ducasp) [reply src/video/VDP.hh]: Last commit fixes this and the original comment

**Comment by ducasp** [comment]
> @ducasp: Thanks a lot for your patch!
> I didn't have time yet to review it in detail (and I'm afraid I also won't have time in the following days). But I did have a quick look during lunch break. (Because I only took a quick look my following comments may be completely wrong).
> 
> Seems like a pretty straight-forward implementation. But I'm wondering about two things:
> 
> 1. Is it needed to explicitly store the blinkStateMatrix[] array? Or could it's content be calculated on demand in the getEvenOddMask() function? At first sight i _think_ it should be sufficient to store a reference line number (e.g. some line at which blinkCount==0 and blinkState=false) and from that you can deduce the full content of the array. This new state (either blinkStateMatrix[] or the reference line) should become part of the savestate (via VDP::serialize()).
> 2. After the previous point is addressed, is it still needed to put a synchronization point on every line? Currently it's still needed to call renderer->updateBlinkState(), but I wonder if that's really required, or if it can be handled in a different way (e.g. in SDLRasterizer::drawDisplay() you already query the state per line). The reason why I ask is because frequent synchronization points are relatively expensive to emulate (by itself a sync point or the code that it triggers is not expensive, but it does prevent optimizations in other sub-systems).

@m9710797 :

The main issue is that updateBlinkState currently does nothing, and the renderer currently is the one responsible for "blink" and "EO" related changes, using the EO related call to this function.... If EO and Blink got moved to the pixelrenderer, then probably getEvenOddMask would not be needed. But this is a way larger changer that might require much more time invested.

So, considering that we have the rasterizer checking the vram page when it is presenting, not when image is being built, and that the original approach worked fine as the EO and BLINK (not considering the undocumented feature) occur at most once per frame, that was fine. Once considering the undocumented feature and that blink might occur several times in the same frame, the approach failed to get the desired effect.

I'll check how difficult it would be to have an algorithm calculating it based on the line being requested, instead of having the matrix / synchronization point every frame line, I understand this might be way easier on the CPU.

**Comment by ducasp** [comment]
@m9710797 

I think the latest commit achieves what you have suggested and is faithful to the real effect.

**Comment by m9710797** [comment]
@ducasp : I merged I reworked version of this pull request. See the commit message of 0117dfd2f3da where I mentioned the most important differences.

Thanks again.

**Pull request closed without merging**